docs/en/week09/09-3.md

```diff
@@ -0,0 +1,387 @@
+
+#### Encoder and Decoder
+
+The last layer of encoder has an output of dimension $2d$: first $d$ values refer to means, $\mu$ and remaninig $d$ values are variances $v$. The decoder has a sigmoid activation for last layer to maintain output range $[0,1]$.
+
+```
+d = 20
+
+class VAE(nn.Module):
+    def __init__(self):
+        super().__init__()
+
+        self.encoder = nn.Sequential(
+            nn.Linear(784, d ** 2),
+            nn.ReLU(),
+            nn.Linear(d ** 2, d * 2)
+        )
+
+        self.decoder = nn.Sequential(
+            nn.Linear(d, d ** 2),
+            nn.ReLU(),
+            nn.Linear(d ** 2, 784),
+            nn.Sigmoid(),
+        )
+```
+#### Reparamaterise and forward function
+
+During trainig, the `reparameterise` function is used for the *reparameterization trick*: We cannot backpropagate through the sampler, we simply compute $z=\mu+\epsilon\odot\sqrt{v}$ where $\epsilon \in \mathcal{N}(0,\mathbb{I}_d)$. This allows to flow gradient back to encoder. During test time, we simply use $\mu$ 
+
+```
+ def reparameterise(self, mu, logvar):
+        if self.training:
+            std = logvar.mul(0.5).exp_()
+            epsilon = std.data.new(std.size()).normal_()
+            return eps.mul(std).add_(mu)
+        else:
+            return mu
+
+    def forward(self, x):
+        mu_logvar = self.encoder(x.view(-1, 784)).view(-1, 2, d)
+        mu = mu_logvar[:, 0, :]
+        logvar = mu_logvar[:, 1, :]
+        z = self.reparameterise(mu, logvar)
+        return self.decoder(z), mu, logvar
+```
+
+We use log variance instead of varaince (change of scale) because we want to ensure
+1. Variance is non-negative.
+2. Full range of variance, to make training stable.
+
+Recall the free energy for VAE, 
+$$
+    F(y)=C(y,\tilde{y})+\beta D_{KL}[\mathcal{N}(\mu,v)||\mathcal{N}(0,1)]
+        =C(y,\tilde{y})+\frac{\beta}{2}\sum_{i=1}^{d}v_i-log(v_i)-1+\mu_i^2
+$$
+
+To regualrize the expressivity of the latent, we include KL-divergence between the Gaussian of latent variable and a Normal distribution (\mathcal{N}(0,1)). $C(y,\hat{y})$ represents the reconstruction term. (Also see [Week 8-Practicum]({{site.baseurl}}/images/week8/08-3) for bubble explaination of VAE loss)
+
+Therefore, we define the loss function as 
+```
+def loss_function(x_hat, x, mu, logvar):
+    BCE = nn.functional.binary_cross_entropy(
+        x_hat, x.view(-1, 784), reduction='sum'
+    )
+    KLD = 0.5 * torch.sum(logvar.exp() - logvar - 1 + mu.pow(2))
+
+    return BCE + KLD
+```
+
+Since, VAE is a generative model, we sample from the distribution to generate the following digits:
+
+```
+N=16
+z=torch.rand((N,d)).to(device)
+sample=model.decoder(z)
+```
+
+<center>
+<img src="{{site.baseurl}}/images/week09/09-3/fig_10_cluster_samples.png" height="300px" style="background-color:#DCDCDC;"/><br>
```

Comment:
Reduce the height a little. The aspect ratio is wrong because it's too high.

Reply:
Fixed, this is the only image which has background property associated with, to display the legend on right side.